### PR TITLE
Make elements with no content invisible

### LINKED ISSUE #520: `isVisible()` does not account for elements that have no visible content and therefore no dimensions
`isVisible()` returns `true` for the following cases:

```html
<div></div>
```

```html
<div>
  <span hidden>This text is hidden</span>
</div>
```

### PR DESCRIPTION
**Pull request by Jym77**
Closes #520.

Not sure this is what you had in mind 🤷

### CODE REVIEW

File: packages/alfa-rules/src/common/predicate/is-visible.ts [hunk 1]
```diff
@@ -70,3 +85,18 @@ function isClipped(
...
+  [
+    "audio",
+    "canvas",
+    "embed",
+    "iframe",
+    "img",
+    "input",
+    "object",
+    "video",
+  ].includes(element.name);
```
Comment (kasperisager) [inline packages/alfa-rules/src/common/predicate/is-visible.ts:102]:
```suggestion
const isReplaced = hasName(
  "audio",
  "canvas",
  "embed",
  "iframe",
  "img",
  "input",
  "object",
  "video",
);
```

File: packages/alfa-rules/src/common/predicate/is-visible.ts [hunk 2]
```diff
@@ -70,3 +85,18 @@ function isClipped(
       .some(isClipped);
   };
 }
+
+/**
+ * @see https://html.spec.whatwg.org/multipage/rendering.html#replaced-elements
```
Comment (kasperisager) [inline packages/alfa-rules/src/common/predicate/is-visible.ts:90]:
```suggestion
 * @see https://html.spec.whatwg.org/#replaced-elements
```

File: packages/alfa-rules/src/common/predicate/is-visible.ts [hunk 3]
```diff
@@ -7,6 +7,8 @@ import { Style } from "@siteimprove/alfa-style";
 
 import { isRendered } from "./is-rendered";
 import { isTransparent } from "./is-transparent";
+import { Iterable } from "@siteimprove/alfa-iterable";
+import every = Iterable.every;
```
Comment (kasperisager) [inline packages/alfa-rules/src/common/predicate/is-visible.ts:11]:
```suggestion
const { every } = Iterable;
```

File: packages/alfa-rules/src/common/predicate/is-visible.ts [hunk 4]
```diff
@@ -7,6 +7,8 @@ import { Style } from "@siteimprove/alfa-style";
 
 import { isRendered } from "./is-rendered";
 import { isTransparent } from "./is-transparent";
+import { Iterable } from "@siteimprove/alfa-iterable";
```
Comment (kasperisager) [inline packages/alfa-rules/src/common/predicate/is-visible.ts:10]: This should be moved up to other `@siteimprove/*` imports.

File: packages/alfa-rules/src/common/predicate/is-visible.ts [hunk 5]
```diff
@@ -33,7 +35,20 @@ export function isVisible(device: Device, context?: Context): Predicate<Node> {
...
+    },
+    // non-replaced elements with no visible child are not visible
+    // replaced elements are assumed to be replaced by something visible.
+    not(
+      and(
+        isElement,
+        and(not(isReplaced), (element) =>
+          every(
+            element.children({ nested: true, flattened: true }),
+            not(isVisible(device, context))
```
Comment (kasperisager) [inline packages/alfa-rules/src/common/predicate/is-visible.ts:47]: The complexity of `isVisible()` is currently something along the lines of `O(d)`, with `d` being the depth of the node, as it walks all the way up the ancestor chain. If we evaluate it for every child of the node as well, we'll end up with something along the lines of `O(d * n)`, with `n` being the size of the subtree of the node. That seems... excessive 🙈 We should probably throw a `Cache` in there somewhere to keep the complexity at `O(d)`.

**Review by kasperisager — CHANGES_REQUESTED** [review]

File: packages/alfa-rules/src/common/predicate/is-visible.ts (hunk 1 above)
Comment (Jym77) [reply packages/alfa-rules/src/common/predicate/is-visible.ts:102]: 🤦

File: packages/alfa-rules/src/common/predicate/is-visible.ts (hunk 5 above)
Comment (Jym77) [reply packages/alfa-rules/src/common/predicate/is-visible.ts:47]: Gaaah 🙈

**Comment by Jym77** [comment]
Gaah, `isTransparent` is also calling on parents… 🙈

**Comment by Jym77** [comment]
Cache everything!

File: packages/alfa-rules/src/common/predicate/is-rendered.ts [hunk 6]
```diff
@@ -6,34 +7,41 @@ import { Style } from "@siteimprove/alfa-style";
 
 const { isElement, hasName } = Element;
 
+const cache = Cache.empty<[Device, Context, Node], boolean>();
```
Comment (kasperisager) [inline packages/alfa-rules/src/common/predicate/is-rendered.ts:10]:
```suggestion
const cache = Cache.empty<Device, Cache<Context, Cache<Node, boolean>>>();
```

The `Cache` class uses reference equality, not structural equality, in order to be able to only hold weak references to object types.

**Review by kasperisager — CHANGES_REQUESTED** [review]

File: packages/alfa-rules/src/common/predicate/is-rendered.ts (hunk 6 above)
Comment (Jym77) [reply packages/alfa-rules/src/common/predicate/is-rendered.ts:10]: Yeah, I actually thought so yesterday night 🙈

**Review by kasperisager — CHANGES_REQUESTED** [review]

**Review by kasperisager — APPROVED** [review]

**Pull request merged**